plugins/kotlin-dataframe/src/org/jetbrains/kotlinx/dataframe/plugin/impl/api/groupBy.kt

```diff
@@ -199,4 +210,217 @@ abstract class GroupByAggregator(val defaultName: String) : AbstractSchemaModifi
+    val Arguments.receiver by groupBy()
+    val Arguments.resultName: String? by arg(defaultValue = Present(null))
+    val Arguments.expression by type()
+
+    override fun Arguments.interpret(): PluginDataFrameSchema {
+        val aggregated = makeNullable(simpleColumnOf(resultName ?: defaultName, expression.type))
+        return PluginDataFrameSchema(receiver.keys.columns() + aggregated)
+    }
+}
+
+/** Implementation for `sumOf` */
+class GroupBySumOf : GroupByAggregatorExpressionSum(defaultName = "sum")
+
+/**
+ * Provides a base implementation for a custom schema modification interpreter
+ * that groups data by specified criteria and produces aggregated results.
+ *
+ * The class uses a `defaultName` to define a fallback name for the result column
+ * if no specific name is provided. It leverages `Arguments` properties to define
+ * and resolve the group-by receiver, result name, and expression type.
+ *
+ * Key Components:
+ * - `receiver`: Represents the input data that will be grouped.
+ * - `resultName`: Optional name for the resulting aggregated column. Defaults to `defaultName`.
+ * - `columns`: ColumnsResolver to define which columns to include in the grouping operation.
+ */
+abstract class GroupByAggregator3(val defaultName: String) : AbstractSchemaModificationInterpreter() {
+    val Arguments.receiver by groupBy()
+    val Arguments.name: String? by arg(defaultValue = Present(null))
+    val Arguments.columns: ColumnsResolver? by arg()
+
+    override fun Arguments.interpret(): PluginDataFrameSchema {
+        if (name == null) {
+            val resolvedColumns = columns?.resolve(receiver.keys)?.map { it.column }!!.toList()
+            return PluginDataFrameSchema(receiver.keys.columns() + resolvedColumns)
+        } else {
+            val resolvedColumns = columns?.resolve(receiver.keys)?.map { it.column }!!.toList()
+            // TODO: how to handle type of multiple columns
+            val aggregated =
+                makeNullable(simpleColumnOf(name ?: defaultName, (resolvedColumns[0] as SimpleDataColumn).type.type))
+            return PluginDataFrameSchema(receiver.keys.columns() + aggregated)
+        }
+    }
+}
+
+/** Implementation for `sum` */
+class GroupBySum0 : GroupByAggregator3(defaultName = "sum")
+
+/** Implementation for `median` */
+class GroupByMedian0 : GroupByAggregator3(defaultName = "median")
+
+/** Implementation for `median` */
+class GroupByMin0 : GroupByAggregator3(defaultName = "min")
+
+/** Implementation for `median` */
+class GroupByMax0 : GroupByAggregator3(defaultName = "max")
+
+abstract class GroupByAggregatorMean(val defaultName: String) : AbstractSchemaModificationInterpreter() {
+    val Arguments.receiver by groupBy()
+    val Arguments.name: String? by arg(defaultValue = Present(null))
+    val Arguments.columns: ColumnsResolver? by arg()
+
+    override fun Arguments.interpret(): PluginDataFrameSchema {
+        if (name == null) {
```

Comment:
Missing a test case here:
`personsDf.groupBy { city }.mean { age and yearsToRetirement }.compareSchemas()`

Runtime:
city: String
mean: Double

Compile:
age: Double
yearsToRetirement: Double
city: String

Reply:
It seems that we agreed that support of multiple columns is under the question?

Reply:
True, at least until after the statistics types are set in stone. That said, for `mean`, the result will always be `Double` :)